src/v/cluster/rm_stm.cc

```diff
@@ -2830,4 +2856,118 @@ ss::future<> rm_stm::clear_old_idempotent_pids() {
+      reqs.tail_seq,
+      reqs.term,
+      reqs.cache.size());
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const rm_stm::mem_state& state) {
+    fmt::print(
+      o,
+      "{{ estimated: {}, tx_start: {}, tx_starts: {}, expected: {}, preparing: "
+      "{} }}",
+      state.estimated.size(),
+      state.tx_start.size(),
+      state.tx_starts.size(),
+      state.expected.size(),
+      state.preparing.size());
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const rm_stm::log_state& state) {
+    fmt::print(
+      o,
+      "{{ fence_epochs: {}, ongoing_m: {}, ongoing_set: {}, prepared: {}, "
+      "aborted: {}, abort_indexes: {}, seq_table: {}, tx_seqs: {}, expiration: "
+      "{}}}",
+      state.fence_pid_epoch.size(),
+      state.ongoing_map.size(),
+      state.ongoing_set.size(),
+      state.prepared.size(),
+      state.aborted.size(),
+      state.abort_indexes.size(),
+      state.seq_table.size(),
+      state.tx_seqs.size(),
+      state.expiration.size());
+    return o;
+}
+
+void rm_stm::setup_metrics() {
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+    namespace sm = ss::metrics;
+    auto ns_label = sm::label("namespace");
+    auto topic_label = sm::label("topic");
+    auto partition_label = sm::label("partition");
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    const auto& ntp = _c->ntp();
+    const std::vector<sm::label_instance> labels = {
+      ns_label(ntp.ns()),
+      topic_label(ntp.tp.topic()),
+      partition_label(ntp.tp.partition()),
+    };
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("tx:partition"),
+      {
+        sm::make_gauge(
+          "idempotency_num_pids_inflight",
+          [this] { return _inflight_requests.size(); },
+          sm::description(
+            "Number of pids with in flight idempotent produce requests."),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "tx_num_inflight_requests",
+          [this] { return _log_state.ongoing_map.size(); },
+          sm::description("Number of ongoing transactional requests."),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_gauge(
+          "tx_mem_tracker_consumption_bytes",
+          [this] { return _tx_root_tracker.consumption(); },
+          sm::description("Total memory bytes in use by tx subsystem."),
+          labels)
+          .aggregate(aggregate_labels),
+      });
+}
+
+ss::future<> rm_stm::maybe_log_tx_stats() {
+    if (likely(!_ctx_log.logger().is_enabled(ss::log_level::debug))) {
+        co_return;
+    }
+    auto units = co_await _state_lock.hold_read_lock();
+    _ctx_log.debug("tx mem tracker breakdown: {}", _tx_root_tracker.debug());
```

Comment:
Based on recent conversations it looks like we want to avoid multi-line log messages? Maybe we can have `debug()` return a JSON encoded string of the memory tree instead?

Reply:
Yep, done.